### PR TITLE
[GC] Partition and aggregate expired addresses more efficiently

### LINKED ISSUE #5612: [GC] Improve performance of `minus` on very large repositories
GC is too slow on large repositories.  Improve performance of `minus`:

* Persist read ranges
* Dedupe range IDs

  This happens in current code, but should be retained in the rewrite for
  the next section!
* Perform RDD-level operations on addresses

  1. Use aggregateByKey exclusively to allow complete control over
     partitioning and operations.
  1. For addresses, aggregateByKey using a HashPartitioner.

     While a RangePartitioner is more commonly used, the quality of the
     partition depends on Spark being able to compute a good "sketch" of the
     range.  Addresses are heavily skewed in practice, many addresses appear
     in very few ranges and some appear an enormous number of times.  And
     the number of occurrences of an address depends on its associated path
     in lakeFS branches, so it is heavily correlated with individual ranges.
     This makes RangePartitioner give poor distributions.  A HashPartitioner
     mixes all addresses together and gives much distributions in practice,
     leading to faster execution times on real data.

A draft code exists that performs these and makes a large repository
somewhat more manageable.

### PR DESCRIPTION
**Pull request by arielshaqed**
New flow:

* Work on RDDs that hold expired and kept ranges/addresses together
* Remove range duplicates
* Compute a reasonable number of range IDs to expand per partition.

  New configuration option:
  `lakefs.gc.address.approx_num_ranges_to_spread_per_partition`
* Expand ranges and hash-partition

  We hash-partition rather than range-partition.  Range partitions work well
  (give similar sizes) when Spark can compute a good "sketch" of the ranges.
  However we need to reduce over object addresses.  Object addresses are
  random _and_ appear with a very unstable distribution.  So there is no
  good way for Spark to find a good range partition.  Hash partitions can
  have their sizes skewed by around $3\cdot\sqrt N$; even for N=1000
  partitions this is very acceptable.
* Directly aggregateByKey to reduce addresses and find those that appear
  only in expired commits.

Also close and delete ranges and metaranges when we are done with them.
This is surprisingly difficult to do correctly in a garbage-collected
language, because it is hard to track lifetimes.  In particular, iterators
extend lifetime beyond dynamic scope.[^1]

Fixes #5612.

[^1]: It is embarrassingly easy to do this in a language with explicit
    lifetimes, such as C++.  And also faster.

### CODE REVIEW

**Comment by arielshaqed** [comment]
**Reviewer:** Please pay special attention to bugfix 4e3a5682dd0fcebbd4eb5dc62c37eb213a4a272b.  This is a mistake that we have managed to make on _every rewrite_ of the code.  (_Luckily this time around we have tests, but it does indicate again that caution is required when writing and reviewing this code!_)

File: clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala [hunk 1]
```diff
@@ -103,90 +105,113 @@ class GarbageCollector(val rangeGetter: RangeGetter) extends Serializable {
...
+      .aggregateByKey(0, numRangePartitions)(
+        (u: Int, expire: Boolean) => u | (if (expire) 1 else 2),
+        (a, b: Int) => a | b
+      )
+      .map({ case (r, expireInt) =>
+        // Ranges appearing only as expire should expire.  Ranges appearing
+        // only as keep should be kept.  Ranges appearing as both expire and
+        // keep should still be kept - to protect their contents from
+        // expired ranges that intersect.
+        (r, expireInt == 1)
```
Comment (Jonathan-Rosenberg) [inline clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala:196]:
```suggestion
      .aggregateByKey(true, numRangePartitions)(
        (expireAcc: Boolean, expire: Boolean) => expireAcc && expire,
        (expireAcc1, expireAcc2: Int) => expireAcc1 && expireAcc2
      )
      .collect()
```
?

I probably don't understand something...

File: clients/spark/core/src/main/scala/io/treeverse/clients/LakeFSContext.scala [hunk 2]
```diff
@@ -64,9 +64,12 @@ object LakeFSContext {
   val LAKEFS_CONF_GC_NUM_COMMIT_PARTITIONS = "lakefs.gc.commit.num_partitions"
   val LAKEFS_CONF_GC_NUM_RANGE_PARTITIONS = "lakefs.gc.range.num_partitions"
   val LAKEFS_CONF_GC_NUM_ADDRESS_PARTITIONS = "lakefs.gc.address.num_partitions"
+  val LAKEFS_CONF_GC_APPROX_NUM_RANGES_TO_SPREAD_PER_PARTITION =
+    "lakefs.gc.address.approx_num_ranges_to_spread_per_partition"
```
Comment (Jonathan-Rosenberg) [inline clients/spark/core/src/main/scala/io/treeverse/clients/LakeFSContext.scala:68]:
```suggestion
  val LAKEFS_CONF_GC_APPROX_NUM_RANGES_PER_PARTITION =
    "lakefs.gc.address.approx_num_ranges_per_partition"
```

File: clients/spark/core/src/main/scala/io/treeverse/clients/SSTableReader.scala [hunk 3]
```diff
@@ -66,29 +73,45 @@ object SSTableReader {
...
   private val reader = new BlockReadableFile(fp)
 
-  def this(sstableFilename: String, companion: GeneratedMessageCompanion[Proto]) =
-    this(new java.io.File(sstableFilename), companion)
+  def this(sstableFilename: String, companion: GeneratedMessageCompanion[Proto], dontOwn: Boolean) =
+    this(new java.io.File(sstableFilename), companion, dontOwn)
 
   def close(): Unit = {
     fp.close()
+    if (!dontOwn) {
```
Comment (Jonathan-Rosenberg) [inline clients/spark/core/src/main/scala/io/treeverse/clients/SSTableReader.scala:112]: why so negative?
can we change it to `own` in all places?

**Review by Jonathan-Rosenberg — APPROVED** [review]
AWESOME, thanks!
I think that more declarative naming and extraction of functions to different modules would make our code much clearer... 
We can definitely use longer names for variables (this is not a comment specifically for this PR but a more general one for the code as a whole).

File: clients/spark/core/src/main/scala/io/treeverse/clients/SSTableReader.scala (hunk 3 above)
Comment (arielshaqed) [reply clients/spark/core/src/main/scala/io/treeverse/clients/SSTableReader.scala]: Fine, be positive like that.

"Think like a proton, always positive" motivational image

Done.

**Review by arielshaqed — COMMENTED** [review]
Thanks!

**Comment by arielshaqed** [comment]
Thanks!

Pulling after doing all the easy bits you requested :-)

**Pull request merged**